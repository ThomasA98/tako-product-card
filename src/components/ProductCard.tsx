import React, { createContext } from 'react'
import { useProduct } from '../hooks/useProduct'

import { onChangeArgs, Product, ProductCardHandlers, ProductCardInitialValues, ProductContextExhibitor } from '../interfaces/interfaces'

import styles from '../styles/styles.module.css'

export const ProductContext = createContext<ProductContextExhibitor>({} as ProductContextExhibitor)
const { Provider } = ProductContext

export interface ProductCardProps {
    product: Product
    children: (args: ProductCardHandlers) => JSX.Element
    className?: string
    style?: React.CSSProperties
    onChange?: (args: onChangeArgs) => void
    value?: number
    initialValues?: ProductCardInitialValues
}

export const ProductCard: React.FC<ProductCardProps> = ({ children, product, className = '', style, onChange, value, initialValues }) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues })

    return (
        <Provider value={{ product, counter, increaseBy, maxCount }}>
            <div className={`${styles.productCard} ${className}`} style={ style } >
                { children({
                    count: counter,
                    product,
                    maxCount,
                    isMaxCountReached,
                    increaseBy,
                    reset,
                }) }
            </div>
        </Provider>
    )
}