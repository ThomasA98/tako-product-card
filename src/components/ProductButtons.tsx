import React, { useCallback, useContext } from 'react'
import { ProductContext } from './ProductCard'

import styles from '../styles/styles.module.css'

export interface ProductButtonsProps {
    className?: string
    style?: React.CSSProperties
}

export const ProductButtons: React.FC<ProductButtonsProps> = ({ className = '', style }) => {

    const { counter, increaseBy, maxCount } = useContext(ProductContext)

    const isMaxReached = useCallback(() => counter === maxCount, [ counter, maxCount ])

    return (
        <div className={ `${styles.buttonsContainer} ${className}` } style={ style }>
            <button onClick={ increaseBy(-1) } className={ styles.buttonMinus }>-</button>

            <div className={ styles.countLabel } >{ counter }</div>

            <button onClick={ increaseBy(1) } className={ `${styles.buttonAdd} ${isMaxReached() ? styles.disabled : ''}` } disabled={isMaxReached()}>+</button>
        </div>
    )
}