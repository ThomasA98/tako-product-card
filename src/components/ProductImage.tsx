import React, { useContext } from 'react'
import { ProductContext } from './ProductCard'

import styles from '../styles/styles.module.css'
import noImage from '../assets/no-image.jpg'

export interface ProductImageProps {
    image?: string
    className?: string
    style?: React.CSSProperties
}

export const ProductImage: React.FC<ProductImageProps> = ({ image, className = '', style }) => {

    const { product } = useContext(ProductContext)

    const imgToShow = image ?? product.image ?? noImage

    return(
        <img className={ `${styles.productImg} ${className}` } style={ style } src={ imgToShow } alt="Product Image" />
    )
}