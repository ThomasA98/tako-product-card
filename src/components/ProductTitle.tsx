import React, { useContext } from 'react'
import { ProductContext } from './ProductCard'

import styles from '../styles/styles.module.css'

export interface ProductTitleProps {
    title?: string
    className?: string
    style?: React.CSSProperties
}

export const ProductTitle: React.FC<ProductTitleProps> = ({ title, className = '', style }) => {

    const { product } = useContext(ProductContext)

    const titleShow = title ?? product.title

    return (
        <span className={ `${styles.productDescription} ${className}` } style={ style } >{ titleShow }</span>
    )
}