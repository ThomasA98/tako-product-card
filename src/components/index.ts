import { ProductCard as ProductCardHOC } from './ProductCard'

import { ProductTitle } from './ProductTitle'
import { ProductImage } from './ProductImage'
import { ProductButtons } from './ProductButtons'

export interface ProductCardHOCProps {
    Title   : typeof ProductTitle
    Image   : typeof ProductImage
    Buttons : typeof ProductButtons
}

export const ProductCard = Object.assign<typeof ProductCardHOC, ProductCardHOCProps>(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons,
})

export { ProductTitle } from './ProductTitle'
export { ProductImage } from './ProductImage'
export { ProductButtons } from './ProductButtons'

export default ProductCard