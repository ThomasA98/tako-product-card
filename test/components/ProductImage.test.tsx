import React from 'react'
import renderer from 'react-test-renderer'

import { ProductImage, ProductCard } from '../../src/components'

import { product1 } from '../data/productos'

describe('ProductTitle', () => {

    test('Debe de mostrar la imagen', () => {

        const wrapper = renderer.create(
            <ProductImage image={ product1.image } />
        )

        expect(wrapper.toJSON()).toMatchSnapshot()

    })

    test('Debe de mostrar la imagen del producto', () => {

        const wrapper = renderer.create(
            <ProductCard product={product1} >
                {() => (
                    <ProductImage />
                )}
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot()

    })

})