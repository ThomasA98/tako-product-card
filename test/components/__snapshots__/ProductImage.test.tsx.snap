// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle Debe de mostrar la imagen 1`] = `
<img
  alt="Product Image"
  className="productImg "
  src="./coffee-mug.png"
/>
`;

exports[`ProductTitle Debe de mostrar la imagen del producto 1`] = `
<div
  className="productCard "
>
  <img
    alt="Product Image"
    className="productImg "
    src="./coffee-mug.png"
  />
</div>
`;
